0000-crypto-construct-language.md

```diff
@@ -0,0 +1,761 @@
+- Feature Name: crypto_construct_language
+- Start Date: 2020-01-29
+- RFC PR:
+- Ursa Issue:
+- Version: 0
+
+# Summary
+[summary]: #summary
+
+This RFC describes a new feature, inspired by Bitcoin script, for serializing
+all cryptographic constructs as a series of data and command tokens that
+describe the use/application of the construct. For instance, a secret key would
+be serialized as an encrypted key with commands for unwrapping the key. Anybody
+wanting to access the secret key can execute the script and if they provide the
+proper inputs, the result will be the unwrapped key. This applies for all
+cryptographic constructs from simple constructs such as key storage up to the
+most complex such as multi-factor key rotation schemes.
+
+# Motivation
+[motivation]: #motivation
+
+By storing cryptographic constructs in "functional" form, we abstract away the
+underlying cryptographic privitive/operation providers as well as provide a
+standardized language for describing arbitrarily complex cryptogrphic
+constructs in serialized form (e.g. DID document, blockchain transaction
+records, encrypted packets, etc).
+
+This came from the DID document standardization effort when we ran into the
+need for storing complex cryptographic constructs for different controller
+binding and key rotation operations. Those operations typically have one or
+more ways that a controller can prove control over an identity and/or force
+a rotation away from a compromised key. This RFC seeks to standardize the
+language we use to describe those constructs so that regarless of the
+underlying crypto library, the construct can be recreated and the operation
+executed.
+
+This also came from the did:git DID method spec effort where we ran into the
+need for describing cryptographic constructs for enforcing governance rules
+for a repo. An example rule would be: changes to the MAINTAINERS file requires
+a commit that is signed by at least 2 of the maintainers of the project. The
+rules need to be serialized in a form that is machine parsable and machine
+executable with context specific references to files and commits to read data
+used in the construct creation. This RFC again seeks to standardize the
+language we use to describe those rules.
```

Comment:
You mention these are for describing rules, but in the examples below you use hard-coded values for keys and signatures (implying a specific execution stream that will be interpreted and executed by an interpreter). Which is it? Both?

Maybe you can show such an example rule for MAINTAINERS file changes encoded in CCL.

Reply:
addressed.